pennylane/capture/meta_type.py

```diff
@@ -0,0 +1,153 @@
+            return qml.s_prod(b, a)
+
+        @staticmethod
+        def _rmul(a, b):
+            return qml.s_prod(b, a)
+
+        @staticmethod
+        def _add(a, b):
+            return qml.sum(a, b)
+
+    jax.core.raise_to_shaped_mappings[AbstractOperator] = lambda aval, _: aval
+
+    return AbstractOperator
+
+
+class PLXPRMeta(abc.ABCMeta):
+    """A metatype that:
+
+    * automatically registers a jax primitive to ``cls._primitive``
+    * Dispatches class creation to ``cls._primitive.bind`` instead of normal class
+    creation when the primitive is defined and plxpr capture is enabled.
+
+    See ``pennylane/capture/explanations.md`` for more detailed information on how this technically
+    works.
+
+    """
+
+    def __init__(cls, *args, **kwargs):
+
+        super().__init__(cls, args, kwargs)
+
+        # Called when constructing a new type that has this metaclass.
+        # Similar to __init_subclass__ , this allows us to run this code
+        # every time we define a new class
+
+        if not has_jax:
+            cls._primitive = None
+            return
+
+        cls._primitive = jax.core.Primitive(cls.__name__)
+
+        @cls._primitive.def_impl
+        def _(*args, **kwargs):
+            if "n_wires" not in kwargs:
+                return type.__call__(cls, *args, **kwargs)
+            n_wires = kwargs.pop("n_wires")
+            wires = args[-n_wires:]
+            args = args[:-n_wires]
+            return type.__call__(cls, *args, wires=wires, **kwargs)
+
+        # logic here will be extended when we make more things use this meta class
+        abstract_type = _get_abstract_operator()
+
+        @cls._primitive.def_abstract_eval
+        def _(*_, **__):
+            return abstract_type()
+
+    def _primitive_bind_call(cls, *args, **kwargs):
+        if cls._primitive is None:
+            # guard against this being called when primitive is not defined.
+            return type.__call__(cls, *args, **kwargs)
+
+        iterable_wires_types = (list, tuple, qml.wires.Wires, range)
+        if "wires" in kwargs:
+            wires = kwargs.pop("wires")
+            wires = tuple(wires) if isinstance(wires, iterable_wires_types) else (wires,)
+            kwargs["n_wires"] = len(wires)
+            args += wires
+        elif args and isinstance(args[-1], iterable_wires_types):
+            kwargs["n_wires"] = len(args[-1])
+            args = args[:-1] + tuple(args[-1])
+        else:
+            kwargs["n_wires"] = 1
+        return cls._primitive.bind(*args, **kwargs)
+
+    def __call__(cls, *args, **kwargs):
+        # this method is called everytime we want to create an instance of the class.
+        # default behavior uses __new__ then __init__
+        # when tracing is enabled, we want to
+
+        if not plxpr_enabled():
+            return type.__call__(cls, *args, **kwargs)
+        # use bind to construct the class if we want class construction to add it to the jaxpr
```

Comment:
```suggestion

        if not plxpr_enabled():
            return type.__call__(cls, *args, **kwargs)
        # when tracing is enabled, we want to
        # use bind to construct the class if we want class construction to add it to the jaxpr
```

Reply:
yeah.... sometimes I get distracted trying to figure out what words to used 🤦